Improve CurrencyEditText

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #5906
Depends on Fluxc PR https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2296
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR tries to improve our `CurrencyEditText` and its layout `WCMaterialOutlinedCurrencyEditTextView`, the list of improvements added is:
1. Ability to move the cursor and edit content anywhere, previously the changes always occurred at the last position.
2. Ability to differentiate between `0` and empty state, this behavior is optional through the property `supportsEmptyState`.
3. Support of negative values, this behavior is optional via the property `supportsNegativeValues`
4. The currency handling is done now using the `prefix`/`suffix` properties of `TextInputLayout`, which makes the cursor position more accurate, instead of pointing to the currency when currency is on the right.
5. Support of two edit modes: one with full formating, and one without it. This is useful for making the editing feels more natural for some screens.

This PR doesn't change any behavior to the screens that were using the `CurrencyEditText`, meaning both `supportsEmptyState` and `supportsNegativeValues` are set to false, and on future PRs we can update each screen with the appropriate combination.
I also was planning on updating `ProductPricingFragment` to go back to using `WCMaterialOutlinedCurrencyEditTextView`, since it supports negative values now, but I decided to do it in a separate PR since this was is already big, I'll try to work on this on a separate PR, and try to fix this [bug](https://github.com/woocommerce/woocommerce-android/issues/3496), since now we can differentiate between empty states and `0`.

This is how the field behaves now:

**Simple Payments**

https://user-images.githubusercontent.com/1657201/156342883-2432eb36-32db-4b92-8452-89124cfb606a.mov 

**Other fields (without full-formating)**

https://user-images.githubusercontent.com/1657201/156342848-04dff67c-2905-49d6-bad9-9799d0417d52.mov

And this is how the UI compares in before vs after:

| Before | After |
| ----- | ----- |
| <img width="300" alt="Screen Shot 2022-02-24 at 16 59 47" src="https://user-images.githubusercontent.com/1657201/155560718-eb5cbb89-e708-4abf-bed1-f7764b7dc190.png"> | <img width="300" alt="Screen Shot 2022-02-24 at 16 49 43" src="https://user-images.githubusercontent.com/1657201/155560748-fd8828e9-9f5c-43e7-8d4d-84e450734e6b.png"> |
| <img width="300" alt="Screen Shot 2022-02-24 at 17 00 06" src="https://user-images.githubusercontent.com/1657201/155560833-0d731b0c-7405-4cb6-b348-f4c3ac7fca14.png"> | <img width="300" alt="Screen Shot 2022-02-24 at 16 49 11" src="https://user-images.githubusercontent.com/1657201/155560861-4753bdf4-957a-4177-bae0-3b2133a273ea.png"> |
| <img width="300" alt="Screen Shot 2022-02-24 at 16 59 13" src="https://user-images.githubusercontent.com/1657201/155561100-59b08667-50a9-4968-bc48-6bf65fea9302.png"> | <img width="300" alt="Screen Shot 2022-02-24 at 16 50 52" src="https://user-images.githubusercontent.com/1657201/155561113-e353a45c-7963-477e-8e9d-6f9db2d7543c.png"> |
| <img width="300" alt="Screen Shot 2022-02-24 at 16 58 59" src="https://user-images.githubusercontent.com/1657201/155561166-ee71b450-a2f7-4c0f-a6dc-9fc13e7b44a2.png"> | <img width="300" alt="Screen Shot 2022-02-24 at 16 50 29" src="https://user-images.githubusercontent.com/1657201/155561179-f8c8e854-0ed2-41eb-a5d5-7cfe977d3155.png">|

### Testing instructions
The `WCMaterialOutlinedCurrencyEditTextView` is currently used in "Simple Payments" and Order Creation (shipping and fees), and it's used in refund by amount (not available in the app now), so please check for any regressions in those areas.
And please if possible, test the changes with these different options:
1. Different keyboards, I tested with Gboard and Samsung Keyboard.
2. Different formatting options from WC Core, especially different options of currency positioning (please note that to force refetch the settings, you need to fully restart the app).

A remark: for order creation, please make sure to change the code manually to update `ORDER_CREATION_M2` to `true` to be able to add `shipping` and `fees`, and please note that the `fees` are not saved when added yet.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Show the currency symbol as prefix or suffix

hichamboushaba committed:
Update CurrencyEditText to remove the currency symbol

hichamboushaba committed:
Handle selection position

hichamboushaba committed:
Handle deleting content on backspace

hichamboushaba committed:
Cleanup code of CurrencyEditText

hichamboushaba committed:
Update CurrencyAmountDialog to use the new component

hichamboushaba committed:
Add support for negative and empty states

hichamboushaba committed:
Add support for setting textsize and gravity

hichamboushaba committed:
Migrate simple payments to use WCMaterialOutlinedCurrencyEditTextView

hichamboushaba committed:
Remove unused functions

hichamboushaba committed:
Update WCMaterialOutlinedCurrencyEditTextView to support wrap content

hichamboushaba committed:
Update SimplePaymentsDialog to make text resize on typing

hichamboushaba committed:
Make CurrencyEditText as private class

This is useful to avoid accidental uses outside of WCMaterialOutlinedCurrencyEditTextView

hichamboushaba committed:
Finish support for empty state for the currency edit text

hichamboushaba committed:
Update usages of currency livedata to account for potential null values

hichamboushaba committed:
Fix support for negative values

hichamboushaba committed:
Fix handling of `setValue`

hichamboushaba committed:
Remove outdated comment

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Remove clean tests

hichamboushaba committed:
Fix tests

hichamboushaba committed:
Rename extension

hichamboushaba committed:
Handle clearing value when empty state is not supported

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt:
```diff
@@ -108,4 +164,132 @@ class WCMaterialOutlinedCurrencyEditTextView @JvmOverloads constructor(
+    }
+}
+
+private class CurrencyEditText @JvmOverloads constructor(
```
I made this as private class to avoid any accidental of usage outside of `WCMaterialOutlinedCurrencyEditTextView`

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt:
```diff
@@ -3,87 +3,143 @@ package com.woocommerce.android.widgets
+    val editText: TextInputEditText
+        get() = currencyEditText
+
+    @Inject lateinit var wcStore: WooCommerceStore
+    @Inject lateinit var selectedSite: SelectedSite
```
Please tell me what you think about this approach?
Previously we were depending on calling `initView` with the correct `currency` and `currencyFormatter`, with DI this is done for us automatically, and makes the component more portable and more encapsulated, since the price formatting is a UI logic anyway.
With this change I was able to remove some code that we didn't need anymore from the ViewModels.

0nko replied:
Nice!

hichamboushaba committed:
Set imeOptions

hichamboushaba commented on WooCommerce/src/main/res/layout/dialog_simple_payments.xml:
```diff
@@ -52,17 +52,20 @@
         android:layout_height="wrap_content"
+        app:boxBackgroundMode="none"
         android:gravity="center"
-        android:imeOptions="flagNoExtractUi"
```
This value is being set automatically by WCMaterialOutlinedCurrencyEditTextView

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt:
```diff
@@ -108,4 +164,132 @@ class WCMaterialOutlinedCurrencyEditTextView @JvmOverloads constructor(
         super.dispatchThawSelfOnly(container)
     }
+
+    private enum class EditTextLayoutMode {
+        FILL, WRAP
+    }
```
This enum is just for differentiating between two modes:
1. The default mode where the editText will take all the available space of the inputLayout (FILL)
2. The second mode that I added, and it's where the editText will use `wrap_content` inside the inputLayout, this is useful for Simple Payments UI. (WRAP)

hichamboushaba commented:
@nbradbury @shiki @anitaa1990 I requested a review from the three of you due to your involvement with the CurrencyEditText recently 🙏.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/widgets/WCMaterialOutlinedCurrencyEditTextView.kt:
```diff
@@ -3,87 +3,146 @@ package com.woocommerce.android.widgets
+        siteSettings?.let {
+            val currencySymbol = wcStore.getSiteCurrency(selectedSite.get(), siteSettings.currencyCode)
+            when (siteSettings.currencyPosition) {
+                LEFT, LEFT_SPACE -> prefixText = currencySymbol
+                RIGHT, RIGHT_SPACE -> suffixText = currencySymbol
```
We don't take the `SPACE` property into consideration here, I think it's OK since the currency is a separate view from the content now.

hichamboushaba commented on WooCommerce/src/test/kotlin/com/woocommerce/android/widgets/CurrencyEditTextTest.kt:
```diff
@@ -1,144 +0,0 @@
-import org.junit.Test
-import kotlin.test.assertEquals
-
-class CurrencyEditTextTest {
```
@shiki I removed those tests since the logic has been simplified a lot after extracting the currency symbol to a different view, but please let me know if you think we should keep some of them 🙏

shiki replied:
I added these in the context of “we need to fix this so we can release the beta for Simple Payments ASAP” (#5412). You can imagine that I hacked and skipped a lot of things to do just that. 😄 There were many things that I did not understand about `CurrencyEditText` and I didn't have the luxury of time to refactor and question the past decisions. Hence, I added unit tests just to fix one small problem and have some confidence that I didn't make it worse. 😄 

I agree (and my reviewer, @wzieba, does too) that this class should be refactored at some point. So, thank you for doing just that! 🙇 

TLDR No worries about removing these tests. 😆

shiki commented:
@hichamboushaba I noticed that I get this crash when opening Simple Payments:

```
2022-02-24 16:19:56.151 7071-7071/com.woocommerce.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.woocommerce.android.prealpha, PID: 7071
    android.view.InflateException: Binary XML file line #69 in com.woocommerce.android.prealpha:layout/dialog_simple_payments: Binary XML file line #69 in com.woocommerce.android.prealpha:layout/dialog_simple_payments: Error inflating class com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView
    Caused by: android.view.InflateException: Binary XML file line #69 in com.woocommerce.android.prealpha:layout/dialog_simple_payments: Error inflating class com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView
    Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.view.LayoutInflater.createView(LayoutInflater.java:852)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:1004)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:959)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:1121)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1082)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:680)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:532)
        at androidx.fragment.app.Fragment.onCreateView(Fragment.java:1968)
        at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2995)
        at androidx.fragment.app.DialogFragment.performCreateView(DialogFragment.java:492)
        at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:523)
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:261)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:1840)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1764)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1701)
        at androidx.fragment.app.FragmentManager$4.run(FragmentManager.java:488)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.ClassCastException: android.content.res.TypedArray cannot be cast to java.lang.AutoCloseable
        at com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView.<init>(WCMaterialOutlinedCurrencyEditTextView.kt:66)
        at com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView.<init>(WCMaterialOutlinedCurrencyEditTextView.kt:35)
        at com.woocommerce.android.widgets.WCMaterialOutlinedCurrencyEditTextView.<init>(Unknown Source:11)
        at java.lang.reflect.Constructor.newInstance0(Native Method) 
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343) 
        at android.view.LayoutInflater.createView(LayoutInflater.java:852) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:1004) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:959) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:1121) 
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1082) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:680) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:532) 
        at androidx.fragment.app.Fragment.onCreateView(Fragment.java:1968) 
        at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2995) 
        at androidx.fragment.app.DialogFragment.performCreateView(DialogFragment.java:492) 
        at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:523) 
        at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:261) 
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:1840) 
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1764) 
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1701) 
        at androidx.fragment.app.FragmentManager$4.run(FragmentManager.java:488) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7660) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
```

Is this a build problem?

hichamboushaba committed:
Add import of correct extension to avoid crash pre-API 31

hichamboushaba commented:
> @hichamboushaba I noticed that I get this crash when opening Simple Payments:

Thanks for finding this @shiki, this should be fixed now by b024af6.
The cause of the crash is that the type `TypedArray` was [updated](https://developer.android.com/reference/android/content/res/TypedArray#close()) to implement `AutoCloseable` in API 31, so the IDE imported the standard `use` extension that supported `AutoCloseable` interface, while on our case we need the `use` extension that `androidx` adds, I mean `androidx.core.content.res.use`, I fixed the import on the last commit.

hichamboushaba committed:
Force using Locale.ROOT when formatting prices

This would help avoid issue with third party keyboards and the clean regex, check https://github.com/woocommerce/woocommerce-android/issues/3651

hichamboushaba commented:
I'm changing this back to draft, I'm still not super satisfied with the experience, and I want to spend a bit more time checking if there is anything left to improve before moving it back to review.
But you are free to test the branch and give me any early feedback 🙏

shiki commented:
> The cause of the crash is that the type TypedArray was [updated](https://developer.android.com/reference/android/content/res/TypedArray#close()) to implement AutoCloseable in API 31, so the IDE imported the standard use extension that supported AutoCloseable interface, while on our case we need the use extension that androidx adds, I mean androidx.core.content.res.use, I fixed the import on the last commit.

I can confirm that the issue is fixed!

Thanks for explaining the issue. It's amazing that the compilation succeeded when in the end it's just gonna crash. 🤷 

I'll test this again when you think it's ready. 🙂

hichamboushaba committed:
Move cursor to last position if we add minus to a zeroed field

hichamboushaba commented:
I think I'm just overthinking this, I tried a totally different approach using `InputFilter` which wouldn't have the formatting as we have now, but would allow a more free and natural editing experience, but I found out that our beloved Android platform has still a lot of issues, and one of them that it need hackish ways to support proper decimal separators (check this [question](https://stackoverflow.com/questions/3821539/decimal-separator-comma-with-numberdecimal-inputtype-in-edittext/45384821#45384821)), which is doable, but made the whole solution as not worth it, as I was aiming for a simpler solution (you can check the code I was working on [here](https://github.com/woocommerce/woocommerce-android/commit/38529b8c1bb34caf41d895aeaa96bef3ce7419e8))

So let's mark this as ready for review as it's, and we can discuss it as a team, and see if it offers a good UX or not.

Just FYI, the commit b106eb60bcdcaf7048b6d9ba5e761c6650efc07d is aiming at restoring the original regex that we had before this [PR](https://github.com/woocommerce/woocommerce-android/pull/3589), as it's suspected that this might be the cause of the [issue](https://github.com/woocommerce/woocommerce-android/issues/3651), and generally Android uses the standard digits when editing numbers, so it doesn't feel natural to have "Arabic-hindu numerals" there.

hichamboushaba commented:
I think I can't help overthinking this 😅, I installed a build using the alternative solution I started and it started growing on me, the editing feels a bit more natural since I can decide whether I want the decimal separator or not.
But at the same time, I feel like the full formatting offered by the current solution is nice too.

If it's not much, I would love your opinion too before going forward with one of the solutions, I uploaded two videos to show how the other solution works in a [draft PR](https://github.com/woocommerce/woocommerce-android/pull/5940), it's not ready for review though.

nbradbury commented:
> I feel like the full formatting offered by the current solution is nice too

I also like the formatting - this is what I would opt for.

shiki commented:
> If it's not much, I would love your opinion too before going forward with one of the solutions, I uploaded two videos to show how the other solution works in a https://github.com/woocommerce/woocommerce-android/pull/5940, it's not ready for review though.

I have mixed feelings. I like the full formatting in the context of the Simple Payments UI. But for other ones, it feels weird to me. I like the simplicity of https://github.com/woocommerce/woocommerce-android/pull/5940 for non-SImple Payments UIs. 

I suspect that there is a history of why we do it this way. Maybe this can be discussed more widely or perhaps involve the design team? 🤷 

Side note: I tested this PR and looks great to me. The only nitpick that I have is when I place the cursor after the decimal separator, I cannot delete it or the number before it. The backspace doesn't do anything. 😅

hichamboushaba committed:
add more formatting options to SiteParameters

hichamboushaba committed:
Using SiteParameters instead of WCSettingsModel

hichamboushaba committed:
Prepare some abstraction layer to allow having multiple implementations

hichamboushaba committed:
Add the RegularCurrencyEditText

hichamboushaba commented:
Thanks everyone for your opinion.

> I have mixed feelings. I like the full formatting in the context of the Simple Payments UI. But for other ones, it feels weird to me. I like the simplicity of https://github.com/woocommerce/woocommerce-android/pull/5940 for non-SImple Payments UIs.

Then why not both 😄, I will create a final PR that allows the consumers to decide whether they want the full-formatting or not.

> The only nitpick that I have is when I place the cursor after the decimal separator, I cannot delete it or the number before it. The backspace doesn't do anything. 

I know about this, it's a downside of having the regex to remove the other characters.
This would be tricky to fix it, and I don't think it's a major thing, why the user would want to put the cursor before the decimal separator that he didn't add 🤔, let's keep things as they are, and I can revisit this in the future.

hichamboushaba committed:
Add a parameter to control whether full-formatting is enabled or false

shiki commented:
> let's keep things as they are, and I can revisit this in the future.

I agree. 🙂

shiki commented:
> Then why not both 😄, I will create a final PR that allows the consumers to decide whether they want the full-formatting or not.

I forgot to comment about this. 

1. My first response: 🤣 
2. Second response: Sounds good to me!

hichamboushaba committed:
Improve functionning of the regular edit text

hichamboushaba committed:
Code cleanup

hichamboushaba committed:
Use full-formatting for simple payments

hichamboushaba committed:
Merge branch 'trunk' into issue/5906-improve-currency-edittext

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/fees/OrderCreationFeeFragment.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/fees/OrderCreationFeeViewModel.kt
#	WooCommerce/src/main/res/layout/fragment_order_creation_edit_fee.xml
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/fees/OrderCreationEditFeeViewModelTest.kt
#	build.gradle

hichamboushaba committed:
Fix fees after the merge commit

hichamboushaba committed:
Bump fluxc hash

hichamboushaba committed:
Remove empty files

hichamboushaba committed:
Fix tests

hichamboushaba committed:
Fix handling of negative values when using regular formatting

hichamboushaba commented:
@shiki @0nko @nbradbury this is now ready for review, I know it's a big PR, sorry for this, the bulk of the logic is in the file `WCMaterialOutlinedCurrencyEditTextView`.

Just some remarks regarding the alternative way for editing, I tried to account for as many cases as possible:
1. It supports the decimal separator configured on the backend and would convert `.` to it if it was entered.
2. When the field doesn't support empty state, entering `0-` will be converted automatically to `-0`, this makes the experience a bit easier, since we can't remove the `0` digit in this case.
3. The leading `0` will be deleted automatically, this is useful also for cases when a non-empty value is not supported, to avoid having something like `012...` when the user starts typing.

Please check the video I added to the description, it shows some of those scenarios.

hichamboushaba committed:
Bump fluxc hash

hichamboushaba committed:
Fix detekt issue

0nko reviewed (approved):
I played around with the app and it looks great, @hichamboushaba! 🏅

nbradbury reviewed (approved):
Looks great to me as well! :shipit:

shiki reviewed (approved):
I'm all good with this too. Thank you!
